source

```
@@ -116611,6 +116678,80 @@ document.body.appendChild(text);
+
+     <li><p><var>Document</var>'s <span data-x="concept-document-allow-declarative-shadow-roots">allow
+     declarative shadow roots</span> is true.</p></li>
+
+     <li><p>The <span>adjusted current node</span> is not the topmost element in the <span>stack of
+     open elements</span>.</p></li>
+    </ul>
+
+    <p>If <var>template start tag</var> is not a <span>declarative shadow root start tag</span>,
+    then <span>insert an HTML element</span> for the token.</p>
+
+    <p>Otherwise, run these steps:</p>
+
+    <ol>
+     <li><p>Let the <var>declarative shadow host element</var> be equal to the <span>adjusted
+     current node</span>.</p></li>
+
+     <li><p><span>Insert a foreign element</span> for <var>template start
+     tag</var>, in the <span>HTML namespace</span> and with <var>only add to element stack</var>
+     set to true, setting <var>template</var> equal to the returned element.</p></li>
+
+     <li><p>Let the <var>declarative shadow mode</var> be the value of <var>template start
+     tag</var>'s "shadowroot" attribute.</p></li>
+
+     <li><p>If <var>template start tag</var> had an attribute with the name
+     "shadowrootdelegatesfocus", then let <var>declarative shadow delegates focus</var> be true.
+     Otherwise let it be false.</p></li>
+
+     <li><p>Run <span data-x="concept-attach-a-shadow-root">attach a shadow root</span> with
+     <var>element</var> equal to <var>declarative shadow host element</var>, <var>mode</var>
+     equal to <var>declarative shadow mode</var>, <var>clonable</var> equal to true,
+     <var>delegatesFocus</var> equal to <var>declarative shadow delegates focus</var>,
+     and <var>slotAssignment</var> equal to "<code data-x="">named</code>".</p></li>
+
+     <li><p>If an exception was thrown by <span data-x="concept-attach-a-shadow-root">attach a
+     shadow root</span>, then catch it, and run these steps:</p>
+
+     <ol>
+      <li><p>Optionally report the exception to a developer console.</p></li>
```

Should this also go to `window.onerror`?

---

Since it's optional, should we do that? Is there a comparable situation elsewhere?

---

Well I would expect it to be required for `window.onerror` if we do that. It's what we do for custom elements as well, no? Seems a bit weird to eat exceptions completely.

---

Good point. Done.

---

I was thinking we'd use the "report the exception" infrastructure here. Again, akin to custom elements. Firing an error event on the host element could also work perhaps, but seems a bit weird.

---

Ok, I can do this also, if you prefer. But I was worried about the fact that ["report the exception"](https://html.spec.whatwg.org/#report-the-exception) has fairly specific text about "for the relevant script" with line numbers and the script's settings object and such. There's no script here, so if I just link to "report the exception", the spec will be kind of broken right? It actually seems like it might already be so for ["invoke custom element reactions"](https://html.spec.whatwg.org/multipage/custom-elements.html#invoke-custom-element-reactions) since the element queue doesn't have (and can't have) associated scripts for each element in the queue, right?

---

Yeah, it's already broken. https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke is another place that invokes it like this.

I don't think we're blocking on resolving that first though. @domenic can correct me.

---

Ok, fair enough! I just made that change.